### ISSUE TITLE
Exception about closed browser in case browser is opened

### ISSUE THREAD

**Issue by andreif2018** [selenide/selenide#862]
## The problem
Hello,

I am getting failed test with following:
 "Java.lang.IIlegalStateException WebDriver has been closed. You need to call open(url) to open browser again".

## Details

As I see the browser window with my url is opened [step 3 attached below], but the test failed with exception.

What can be a solution  to reopen browser and to make assertion in selenide in my case?
Thanks

## Tell us about your environment
* **Selenide Version: 5.0.1**
* **Chrome Version: 70.0.3538.102 For Windows 10 64-bit**
* **Browser Driver Version: ChromeDriver 2.43**
* **Selenium Version: 3.141.5  (Java)**
* **OS Version: Windows 10 64-bit.**

## Code To Reproduce Issue [ Good To Have ]
Parameterized test:
step 1: open(my url);
step 2: close();
step 3: open(my url);
next is my assertion

**Comment by BorisOsipov** [comment]
@andreif2018 Sorry for long response. Are you solve your problem or the issue is still actual?

**Comment by symonk** [comment]
will investigate tomorrow night (not home until then)

**Comment by asolntsev** [comment]
@andreif2018 I guess you call `$` *before* opening a browser. Am I right?

I just tried to call `$` *after* opening a browser, and everything works:

```java
  @Test
  public void issue862() {
    open("http://google.com");
    close();
    open("http://google.com");
    SelenideElement a=$("[name=q]");
    a.shouldBe(visible);     // it works
  }
```

If this is true, you just need to move all `$` calls *after* opening a browser.

**Comment by andreif2018** [comment]
@asolntsev thanks for response, in my case the test is parameterized where parameter is selenide element and I call and assert this parameter after reopen and it does not work. in regular, not parameterized tests it works perfect. what can be solution with selenide? By the way, currently I use workaround in parameterized test by call execute javascript method with "window.close();"

**Comment by asolntsev** [comment]
@andreif2018 Wow. Sorry, but it seems like a total misuse of test parametrization. Don't pass `SelenideElement` as a test parameter. Test parameters are meant to be Strings, Integers or something like that. Usernames, passwords, account numbers etc. You just need to call `$` exactly in the place where you need a web element.

**Comment by symonk** [comment]
ouch @andreif2018 yes please do not iterate in a test with providing different WebElement(SelenideElement) for each iteration, this is a very bad practice.

**Comment by pavel-kolesnikov** [comment]
@asolntsev 
Lets consider this example. For me it looks completely valid. And 2nd call to body.should fails. Is this still not a bug?

```package test;

import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;

import static com.codeborne.selenide.Condition.exist;
import static com.codeborne.selenide.Selenide.$;
import static com.codeborne.selenide.Selenide.open;

public class runrun {
    private static SelenideElement body = $("body");

    public static void main(String[] args) {
        open("https://mail.ru");
        body.should(exist);

        Selenide.close();

        open("https://mail.ru");
        body.should(exist);
    }
}
```

```Starting ChromeDriver 73.0.3683.20 (8e2b610813e167eee3619ac4ce6e42e3ec622017) on port 20114
Only local connections are allowed.
Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.
Feb 07, 2019 1:48:55 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
Feb 07, 2019 1:48:55 PM com.codeborne.selenide.webdriver.BrowserResizer adjustBrowserSize
INFO: Set browser size to 1366x768
Feb 07, 2019 1:48:55 PM com.codeborne.selenide.webdriver.WebDriverFactory logBrowserVersion
INFO: BrowserName=chrome Version=72.0.3626.81 Platform=LINUX
Feb 07, 2019 1:48:55 PM com.codeborne.selenide.webdriver.WebDriverFactory createWebDriver
INFO: Selenide v. 5.1.0
Feb 07, 2019 1:48:55 PM com.codeborne.selenide.webdriver.WebDriverFactory logSeleniumInfo
INFO: Selenium WebDriver v. 3.141.59 build time: 2018-11-14T08:17:03
Feb 07, 2019 1:48:55 PM com.codeborne.selenide.drivercommands.CreateDriverCommand createDriver
INFO: Create webdriver in current thread 1: ChromeDriver -> ChromeDriver: chrome on LINUX (856cacb3e7365de55a9e725ecef2f7e2)
Feb 07, 2019 1:48:57 PM com.codeborne.selenide.drivercommands.CloseDriverCommand run
INFO: Close webdriver: 1 -> ChromeDriver: chrome on LINUX (856cacb3e7365de55a9e725ecef2f7e2)
Feb 07, 2019 1:48:57 PM com.codeborne.selenide.drivercommands.CloseBrowser run
INFO: Trying to close the browser ChromeDriver ...
Feb 07, 2019 1:48:57 PM com.codeborne.selenide.drivercommands.CloseDriverCommand run
INFO: Closed webdriver 1 in 70 ms
Feb 07, 2019 1:48:57 PM com.codeborne.selenide.drivercommands.LazyDriver getAndCheckWebDriver
INFO: No webdriver is bound to current thread: 1 - let's create a new webdriver
Starting ChromeDriver 73.0.3683.20 (8e2b610813e167eee3619ac4ce6e42e3ec622017) on port 11057
Only local connections are allowed.
Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.
Feb 07, 2019 1:48:58 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
Feb 07, 2019 1:48:58 PM com.codeborne.selenide.webdriver.BrowserResizer adjustBrowserSize
INFO: Set browser size to 1366x768
Feb 07, 2019 1:48:58 PM com.codeborne.selenide.webdriver.WebDriverFactory logBrowserVersion
INFO: BrowserName=chrome Version=72.0.3626.81 Platform=LINUX
Feb 07, 2019 1:48:58 PM com.codeborne.selenide.webdriver.WebDriverFactory createWebDriver
INFO: Selenide v. 5.1.0
Feb 07, 2019 1:48:58 PM com.codeborne.selenide.webdriver.WebDriverFactory logSeleniumInfo
INFO: Selenium WebDriver v. 3.141.59 build time: 2018-11-14T08:17:03
Feb 07, 2019 1:48:58 PM com.codeborne.selenide.drivercommands.CreateDriverCommand createDriver
INFO: Create webdriver in current thread 1: ChromeDriver -> ChromeDriver: chrome on LINUX (feb365e155e952a51c9dadd5b70dca0d)
Exception in thread "main" java.lang.IllegalStateException: Webdriver has been closed. You need to call open(url) to open a browser again.
	at com.codeborne.selenide.drivercommands.LazyDriver.getWebDriver(LazyDriver.java:64)
	at com.codeborne.selenide.impl.ElementFinder.getSearchContext(ElementFinder.java:86)
	at com.codeborne.selenide.impl.ElementFinder.getWebElement(ElementFinder.java:74)
	at com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:50)
	at com.codeborne.selenide.commands.Should.should(Should.java:35)
	at com.codeborne.selenide.commands.Should.execute(Should.java:29)
	at com.codeborne.selenide.commands.Should.execute(Should.java:12)
	at com.codeborne.selenide.commands.Commands.execute(Commands.java:144)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:99)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:65)
	at com.sun.proxy.$Proxy0.should(Unknown Source)
	at icerock.runrun.main(runrun.java:20)
Feb 07, 2019 1:48:59 PM com.codeborne.selenide.drivercommands.CloseDriverCommand run
INFO: Close webdriver: 32 -> ChromeDriver: chrome on LINUX (feb365e155e952a51c9dadd5b70dca0d)
```

**Comment by asolntsev** [comment]
@pavel-kolesnikov it seems a very strange design for me. Why do you need a static field? Why do you need to reuse variable for different objects? 

We can argue if it was a good idea, but yes, current Selenide 5.+ behavior prohibits such usage.

**Comment by pavel-kolesnikov** [comment]
Then it seems for me, it should be clearly stated in the documentation, that:

1. neither SelenideElement, not Selenide itself is not threadsafe;
1. SelenideElements must not be used in the static context;
1. SelenideElements  must not be saved anyhow for prolonged use; 
1. SelenideElements should be expected to live no longer than the method they're constructed in.

Do I undesrtand the current contract clearly?

**Comment by pavel-kolesnikov** [comment]
And the actual design for the usage is below. Do i understand it correctly, we're not allowed to do this, and better use "standard" PageObjects, which should be constructed with `new` or `page()`
```
class LoginBlock {
  public static SelenideElement button = $("button");
}

class UserProfileBlock {
  public static SelenideElement logout = $("#logout");
}

class LoginTest {
  @BeforeMethod
  void login(){
    open("url");
    LoginBlock.login.click();
    UserProfileBlock.logout.shouldBe(visible);
  }

  // many tests...
}
```

**Comment by asolntsev** [comment]
@pavel-kolesnikov 

> Do I undesrtand the current contract clearly?
Not exactly, but close. My version is below:

  1. both SelenideElement and Selenide are threadsafe. Every thread gets its own webdriver instance.
  2. SelenideElements can be used in the static context (unless one SelenideElement instance is used in multiple threads concurrently)
  3. SelenideElements  must not be saved anyhow for prolonged use - yes, in some sense it's true (in Selenide 5.x)
  4. SelenideElements should be expected to live no longer than browser is opened.

> And the actual design for the usage is below.

Yes, I think this design is bad. It's a bad practice to have public static fields. Yes, I think it's better to use "standard" PageObjects which should be constructed with `new` or `page()`.

**Comment by pavel-kolesnikov** [comment]
Thank you for the clarification. Will alter my code to be compliant with the current policies.

**Comment by g-one** [comment]
@asolntsev 
I have created all my page objects using **$** as static elements and we havent run all tests as a suite but didnt realize Selenide doesnt support using these static elements in all functions in given page. But as per your earlier comment your saying need to call **$** only where element is needed then how about if the same element need to use by multiple functions in the same class? Now we are in process of refactoring page elements to get through the issues **Java.lang.IIlegalStateException WebDriver has been closed. You need to call open(url) to open browser again**. Though we are using opening the browser for every method using Testng @BeforeMethod but it seems like Selenide trying to find the element before the browser opens when we call new LoginPage() for example.

Any help is much appreciated.

`   static SelenideElement uName = $("#username");
    static SelenideElement pwd = $("#password");
    static SelenideElement loginBtn = $("#login-button");

    public AccountPage login(String userName,String password){
        uName.val(userName);
        pwd.val(password);
        loginBtn.click();
        return page(AccountPage.class);
    }`

**Comment by asolntsev** [comment]
@g-one easy, easy. :)
 fortunately, there is an easy solution. Just remove word `static` from fields. Now your page object should look like this:


```java
private SelenideElement uName = $("#username"); private SelenideElement pwd = $("#password");
private SelenideElement loginBtn = $("#login-button");

public AccountPage login(String userName,String password) {
  uName.val(userName);
  pwd.val(password);
  loginBtn.click();
  return page(AccountPage.class);
}
```

**Comment by pavel-kolesnikov** [comment]
@asolntsev 
> public AccountPage login(String userName,String password) {
should the method be `static`?

**Comment by pavel-kolesnikov** [comment]
nah, stupid comment from me, sorry

**Comment by g-one** [comment]
@asolntsev Thank you, i tried all other options except this :). Seems like $(byText) having issues with new version 5.2.0. Will create log the details in new issue.

**Comment by VladislavShcherba** [comment]
Yes, this is unfortunate stuff.
The idea was to use just `LoginPage.login()` instead of instantiating LoginPage object...
But having SelenideElement-s as static fields is forbid to open/reopen browser during execution of test suite...

**Comment by asolntsev** [comment]
@VladislavShcherba It's not a bad idea to instantiate LoginPage object. Yes, instantiate it. Why not?

See https://www.youtube.com/watch?v=4JJNccWtdNI :)

**Comment by asolntsev** [comment]
@andreif2018  @symonk @pavel-kolesnikov @g-one @VladislavShcherba Congratulations, we fixed this issue in Selenide 5.4.0. Will be released soon.

@pavel-kolesnikov For example, your code from https://github.com/selenide/selenide/issues/862#issuecomment-461374927 will work with Selenide 5.4.0
